File: training-data/scala/LinkedListDemo.scala
class ListNode(val value: Int, var next: Option[ListNode] = None)

object LinkedListDemo {
  def reverseAndCollect(head: Option[ListNode]): List[Int] = {
    var prev: Option[ListNode] = None
    var cur = head
    while (cur.isDefined) {
      val node = cur.get
      val next = node.next
      node.next = prev
      prev = Some(node)
      cur = next
    }
    var result = List[Int]()
    var n = prev
    while (n.isDefined) {
      result = result :+ n.get.value
      n = n.get.next
    }
    result
  }

  def main(args: Array[String]): Unit = {
    val n4 = new ListNode(4)
    val n3 = new ListNode(3, Some(n4))
    val n2 = new ListNode(2, Some(n3))
    val n1 = new ListNode(1, Some(n2))
    println(reverseAndCollect(Some(n1)))
  }
}

File: training-data/objectivec/LinkedListDemo.m
#import <Foundation/Foundation.h>

@interface LLNode : NSObject
@property (nonatomic, strong) NSNumber *value;
@property (nonatomic, strong) LLNode *next;
@end

@implementation LLNode
@end

@interface SinglyLinkedList : NSObject
@property (nonatomic, strong) LLNode *head;
- (void)addValue:(NSNumber *)value;
- (NSArray<NSNumber *> *)toArray;
- (void)reverse;
@end

@implementation SinglyLinkedList
- (void)addValue:(NSNumber *)value {
    LLNode *node = [[LLNode alloc] init];
    node.value = value;
    if (!self.head) {
        self.head = node;
        return;
    }
    LLNode *cur = self.head;
    while (cur.next) cur = cur.next;
    cur.next = node;
}

- (NSArray<NSNumber *> *)toArray {
    NSMutableArray *result = [NSMutableArray array];
    LLNode *cur = self.head;
    while (cur) {
        [result addObject:cur.value];
        cur = cur.next;
    }
    return result;
}

- (void)reverse {
    LLNode *prev = nil;
    LLNode *cur = self.head;
    while (cur) {
        LLNode *next = cur.next;
        cur.next = prev;
        prev = cur;
        cur = next;
    }
    self.head = prev;
}
@end

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        SinglyLinkedList *list = [[SinglyLinkedList alloc] init];
        for (NSInteger i = 1; i <= 5; i++) [list addValue:@(i)];
        NSLog(@"%@", [list toArray]);
        [list reverse];
        NSLog(@"%@", [list toArray]);
    }
    return 0;
}
